### PR TITLE
Improve typing hints

### PR DESCRIPTION
**Pull request by alissonlauffer**
### Proposed changes
 - Change `type1 or type2` to `Union[type1, type2]` because the first one is not a valid syntax for typing hints (though it does not give any SyntaxError).
 - Change `Union[type1, None]` to `Optional[type1]` as proposed in https://github.com/pyrogram/pyrogram/pull/537#pullrequestreview-540450312.

### CODE REVIEW

File: pyrogram/connection/transport/tcp/tcp_abridged.py [hunk 1]
```diff
@@ -41,7 +43,7 @@ async def send(self, data: bytes, *args):
             + data
         )
 
-    async def recv(self, length: int = 0) -> bytes or None:
+    async def recv(self, length: int = 0) -> Union[bytes, None]:
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_abridged.py:46]: Why don't you use `Optional[bytes]`

File: pyrogram/connection/connection.py [hunk 2]
```diff
@@ -19,6 +19,8 @@
 import asyncio
 import logging
 
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/connection/connection.py:22]:
```suggestion
from typing import Optional
```

File: pyrogram/connection/connection.py [hunk 3]
```diff
@@ -79,5 +81,5 @@ async def send(self, data: bytes):
         except Exception:
             raise OSError
 
-    async def recv(self) -> bytes or None:
+    async def recv(self) -> Union[bytes, None]:
```
Comment (KunoiSayami) [inline pyrogram/connection/connection.py:84]:
```suggestion
    async def recv(self) -> Optional[bytes]:
```

File: pyrogram/connection/transport/tcp/tcp_abridged.py [hunk 4]
```diff
@@ -18,6 +18,8 @@
 
 import logging
 
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_abridged.py:21]:
```suggestion
from typing import Optional
```

File: pyrogram/connection/transport/tcp/tcp_abridged.py (hunk 1 above)
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_abridged.py:46]:
```suggestion
    async def recv(self, length: int = 0) -> Optional[bytes]:
```

File: pyrogram/connection/transport/tcp/tcp_abridged_o.py [hunk 5]
```diff
@@ -19,6 +19,8 @@
 import logging
 import os
 
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_abridged_o.py:22]:
```suggestion
from typing import Optional
```

File: pyrogram/connection/transport/tcp/tcp_abridged_o.py [hunk 6]
```diff
@@ -66,7 +68,7 @@ async def send(self, data: bytes, *args):
             )
         )
 
-    async def recv(self, length: int = 0) -> bytes or None:
+    async def recv(self, length: int = 0) -> Union[bytes, None]:
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_abridged_o.py:71]:
```suggestion
    async def recv(self, length: int = 0) -> Optional[bytes]:
```

File: pyrogram/connection/transport/tcp/tcp_full.py [hunk 7]
```diff
@@ -17,6 +17,7 @@
 #  along with Pyrogram.  If not, see <http://www.gnu.org/licenses/>.
 
 import logging
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_full.py:20]:
```suggestion
from typing import Optional
```

File: pyrogram/connection/transport/tcp/tcp_full.py [hunk 8]
```diff
@@ -42,7 +43,7 @@ async def send(self, data: bytes, *args):
 
         await super().send(data)
 
-    async def recv(self, length: int = 0) -> bytes or None:
+    async def recv(self, length: int = 0) -> Union[bytes, None]:
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_full.py:46]:
```suggestion
    async def recv(self, length: int = 0) -> Optional[bytes]:
```

File: pyrogram/connection/transport/tcp/tcp_intermediate.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 #  along with Pyrogram.  If not, see <http://www.gnu.org/licenses/>.
 
 import logging
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_intermediate.py:20]:
```suggestion
from typing import Optional
```

File: pyrogram/connection/transport/tcp/tcp_intermediate.py [hunk 10]
```diff
@@ -35,7 +36,7 @@ async def connect(self, address: tuple):
     async def send(self, data: bytes, *args):
         await super().send(pack("<i", len(data)) + data)
 
-    async def recv(self, length: int = 0) -> bytes or None:
+    async def recv(self, length: int = 0) -> Union[bytes, None]:
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_intermediate.py:39]:
```suggestion
    async def recv(self, length: int = 0) -> Optional[bytes]:
```

File: pyrogram/connection/transport/tcp/tcp_intermediate_o.py [hunk 11]
```diff
@@ -18,6 +18,7 @@
 
 import logging
 import os
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_intermediate_o.py:21]:
```suggestion
from typing import Optional
```

File: pyrogram/connection/transport/tcp/tcp_intermediate_o.py [hunk 12]
```diff
@@ -62,7 +63,7 @@ async def send(self, data: bytes, *args):
             )
         )
 
-    async def recv(self, length: int = 0) -> bytes or None:
+    async def recv(self, length: int = 0) -> Union[bytes, None]:
```
Comment (KunoiSayami) [inline pyrogram/connection/transport/tcp/tcp_intermediate_o.py:66]:
```suggestion
    async def recv(self, length: int = 0) -> Optional[bytes]:
```

File: pyrogram/types/messages_and_media/message_entity.py [hunk 13]
```diff
@@ -16,6 +16,8 @@
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with Pyrogram.  If not, see <http://www.gnu.org/licenses/>.
 
+from typing import Union
```
Comment (KunoiSayami) [inline pyrogram/types/messages_and_media/message_entity.py:19]:
```suggestion
from typing import Optional
```

File: pyrogram/types/messages_and_media/message_entity.py [hunk 14]
```diff
@@ -84,7 +86,7 @@ def __init__(
         self.user = user
 
     @staticmethod
-    def _parse(client, entity, users: dict) -> "MessageEntity" or None:
+    def _parse(client, entity, users: dict) -> Union["MessageEntity", None]:
```
Comment (KunoiSayami) [inline pyrogram/types/messages_and_media/message_entity.py:89]:
```suggestion
    def _parse(client, entity, users: dict) -> Optional["MessageEntity"]:
```

File: pyrogram/types/user_and_chats/user.py [hunk 15]
```diff
@@ -17,7 +17,7 @@
 #  along with Pyrogram.  If not, see <http://www.gnu.org/licenses/>.
 
 import html
-from typing import List
+from typing import List, Union
```
Comment (KunoiSayami) [inline pyrogram/types/user_and_chats/user.py:20]:
```suggestion
from typing import List, Optional
```

File: pyrogram/types/user_and_chats/user.py [hunk 16]
```diff
@@ -201,7 +201,7 @@ def mention(self):
         return Link(f"tg://user?id={self.id}", self.first_name, self._client.parse_mode)
 
     @staticmethod
-    def _parse(client, user: "raw.types.User") -> "User" or None:
+    def _parse(client, user: "raw.types.User") -> Union["User", None]:
```
Comment (KunoiSayami) [inline pyrogram/types/user_and_chats/user.py:204]:
```suggestion
    def _parse(client, user: "raw.types.User") -> Optional["User"]:
```

**Review by KunoiSayami — CHANGES_REQUESTED** [review]
Use `Optional[X]` may more intuitive than `Union[X, None]`
Reference: https://docs.python.org/3/library/typing.html#typing.Optional

**Review by KunoiSayami — APPROVED** [review]

**Review by KunoiSayami — APPROVED** [review]

File: pyrogram/filters.py [hunk 17]
```diff
@@ -824,7 +824,7 @@ class user(Filter, set):
             Defaults to None (no users).
     """
 
-    def __init__(self, users: int or str or list = None):
+    def __init__(self, users: Union[int, str, list] = None):
```
Comment (delivrance) [inline pyrogram/filters.py:827]: This can be improved with `Union[int, str, List[Union[int, str]]]`

File: pyrogram/filters.py [hunk 18]
```diff
@@ -856,7 +856,7 @@ class chat(Filter, set):
             Defaults to None (no chats).
     """
 
-    def __init__(self, chats: int or str or list = None):
+    def __init__(self, chats: Union[int, str, list] = None):
```
Comment (delivrance) [inline pyrogram/filters.py:859]: Same here, `Union[int, str, List[Union[int, str]]]`

**Pull request merged**